### PR TITLE
Add support for wikimedia.org

### PR DESCRIPTION
**Pull request by EhtishamSabir**
<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### Description of your *pull request* and other information

</details>

DESCRIPTION

Fixes #

### CODE REVIEW

**Comment by EhtishamSabir** [comment]
Provide verbose output that clearly demonstrates the problem
 Run your yt-dlp command with -vU flag added (yt-dlp -vU <your command line>)
 Copy the WHOLE output (starting with [debug] Command-line config) and insert it below
Complete Verbose Output

[debug] Command-line config: ['-vU', 'https://commons.wikimedia.org/wiki/File:Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS.webm']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.06.29 [9d339c41e] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: 7ef02a740
[debug] Python 3.8.0 (CPython 64bit) - Windows-10-10.0.19041-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: avconv -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: avprobe -bsfs
[debug] exe versions: none
[debug] Optional libraries: Cryptodome-3.9.9, brotli-1.0.7, certifi-2020.06.20, sqlite3-2.6.0, websockets-9.1
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.06.29, Current version: 2022.06.29
yt-dlp is up to date (2022.06.29)
[debug] [wikimedia.org] Extracting URL: https://commons.wikimedia.org/wiki/File:Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS.webm
[wikimedia.org] Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS.webm: Downloading webpage
[wikimedia.org] Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS.webm: Extracting information
[debug] Default format spec: best/bestvideo+bestaudio
[info] Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS: Downloading 1 format(s): 0
[debug] Invoking http downloader on "https://upload.wikimedia.org/wikipedia/commons/transcoded/d/d7/Die_Temperaturkurve_der_Erde_%28ZDF%2C_Terra_X%29_720p_HD_50FPS.webm/Die_Temperaturkurve_der_Erde_%28ZDF%2C_Terra_X%29_720p_HD_50FPS.webm.480p.vp9.webm"
[debug] File locking is not supported. Proceeding without locking
[download] Destination: Die Temperaturkurve der Erde (ZDF, Terra X) 720p HD 50FPS.webm - Wikimedia Commons [Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS].webm
[download] 100% of 3.16MiB in 00:18

**Comment by pukkandan** [comment]
I see this is copied as-is from #30796. If you want this to be directly merged into yt-dlp (rather than being pulled after upstream merges it), you are going to have to follow our coding conventions

File: yt_dlp/extractor/wikimedia.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
```
Comment (pukkandan) [inline yt_dlp/extractor/wikimedia.py:3]: remove

File: yt_dlp/extractor/_extractors.py [hunk 2]
```diff
@@ -2209,3 +2209,4 @@
 )
 from .zoom import ZoomIE
 from .zype import ZypeIE
+from .wikimedia import WikimediaIE
```
Comment (pukkandan) [inline yt_dlp/extractor/_extractors.py:2212]: add in alphabetical order

File: yt_dlp/extractor/wikimedia.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+from .common import InfoExtractor
+
+import re
+
+from ..utils import (
+    clean_html,
+    determine_ext,
+    get_element_by_class,
+    urljoin,
+    parse_qs as compat_parse_qs,
```
Comment (pukkandan) [inline yt_dlp/extractor/wikimedia.py:13]: don't do this

File: yt_dlp/extractor/wikimedia.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    ExtractorError)
+
+
+class WikimediaIE(InfoExtractor):
+    IE_NAME = 'wikimedia.org'
+    _NETRC_MACHINE = 'wikimediaorg'
```
Comment (pukkandan) [inline yt_dlp/extractor/wikimedia.py:19]: _NETRC_MACHINE must only be set of the extractor has login support

File: yt_dlp/extractor/wikimedia.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            'license': 'md5:62907cddf705a9f7ae7076c15407a977',
+            'author': None, 'subtitles': {'de': [{'ext': 'srt',
+                                                  'url': 'https?://commons.wikimedia.org/w/api.php'}],
+                                          'en-gb': [{'ext': 'srt',
+                                                     'url': 'https?://commons.wikimedia.org/w/api.php'}],
+                                          'nl': [{'ext': 'srt',
+                                                  'url': 'https?://commons.wikimedia.org/w/api.php'}],
+                                          'en': [{'ext': 'srt',
+                                                  'url': 're:https?://commons.wikimedia.org/w/api.php'}]}}
+    }
```
Comment (pukkandan) [inline yt_dlp/extractor/wikimedia.py:38]: This formatting is terrible... Read https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#trailing-parentheses

File: yt_dlp/extractor/wikimedia.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    _VALID_URL = 'https://commons.wikimedia.org/wiki/File:(?P<id>[^/]+)'
+
+    _TEST = {
+        'url': 'https://commons.wikimedia.org/wiki/File:Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS.webm',
+        'info_dict': {
+            'description': 'md5:7cd84f76e7081f1be033d0b155b4a460',
+            'ext': 'webm', 'id': 'Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS',
+            'title': 'Die Temperaturkurve der Erde (ZDF, Terra X) 720p HD 50FPS.webm - Wikimedia Commons',
+            'license': 'md5:62907cddf705a9f7ae7076c15407a977',
+            'author': None, 'subtitles': {'de': [{'ext': 'srt',
```
Comment (pukkandan) [inline yt_dlp/extractor/wikimedia.py:30]: 1. author is not a valid field
2. Fields that return None shouldn't be in tests

File: yt_dlp/extractor/wikimedia.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                                                  'url': 'https?://commons.wikimedia.org/w/api.php'}],
+                                          'en-gb': [{'ext': 'srt',
+                                                     'url': 'https?://commons.wikimedia.org/w/api.php'}],
+                                          'nl': [{'ext': 'srt',
+                                                  'url': 'https?://commons.wikimedia.org/w/api.php'}],
+                                          'en': [{'ext': 'srt',
+                                                  'url': 're:https?://commons.wikimedia.org/w/api.php'}]}}
```
Comment (pukkandan) [inline yt_dlp/extractor/wikimedia.py:37]: This test doesn't pass

File: yt_dlp/extractor/wikimedia.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        ext = determine_ext(url, None)
+        if ext is None:
+            raise ExtractorError('invalid video url', expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/wikimedia.py:44]: I don't think this check is needed. But you want it, do:
```suggestion
        if not determine_ext(url, None):
            raise ExtractorError('The URL does not contain a video', expected=True)
```

Or even better, the valid_url can be improved to not match URLs without extension

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/wikimedia.py (hunk 1 above)
Comment (EhtishamSabir) [reply yt_dlp/extractor/wikimedia.py:3]: thanks for help fixed.  moving to next issue.

File: yt_dlp/extractor/_extractors.py (hunk 2 above)
Comment (EhtishamSabir) [reply yt_dlp/extractor/_extractors.py:2212]: done

File: yt_dlp/extractor/wikimedia.py (hunk 5 above)
Comment (EhtishamSabir) [reply yt_dlp/extractor/wikimedia.py:38]: I tried it to make it look as good as possible please review in next commit

File: yt_dlp/extractor/wikimedia.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    _API_BASE_URL = 'https://commons.wikimedia.org/'
+    _VALID_URL = 'https://commons.wikimedia.org/wiki/File:(?P<id>[^/]+)'
+
+    _TEST = {
+        'url': 'https://commons.wikimedia.org/wiki/File:Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS.webm',
+        'info_dict': {
+            'description': 'md5:7cd84f76e7081f1be033d0b155b4a460',
+            'ext': 'webm', 'id': 'Die_Temperaturkurve_der_Erde_(ZDF,_Terra_X)_720p_HD_50FPS',
+            'title': 'Die Temperaturkurve der Erde (ZDF, Terra X) 720p HD 50FPS.webm - Wikimedia Commons',
+            'license': 'md5:62907cddf705a9f7ae7076c15407a977',
```
[the remark this answers is not in the document]
Comment (EhtishamSabir) [reply yt_dlp/extractor/wikimedia.py:29]: actually license text is too big

File: yt_dlp/extractor/wikimedia.py (hunk 7 above)
Comment (EhtishamSabir) [reply yt_dlp/extractor/wikimedia.py:37]: there was a subtitles issue.

File: yt_dlp/extractor/wikimedia.py (hunk 8 above)
Comment (EhtishamSabir) [reply yt_dlp/extractor/wikimedia.py:44]: I'm using this video extension later in the code but I added your suggestion for Error message

File: yt_dlp/extractor/wikimedia.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        ext = determine_ext(url, None)
+        if ext is None:
+            raise ExtractorError('invalid video url', expected=True)
+        webpage = self._download_webpage(url, video_id)
+        self.report_extraction(video_id)
+        video_url = self._html_search_regex('<source [^>]*src="([^"]+)"', webpage,
+                                            'video URL')
+        license = get_element_by_class('layouttemplate licensetpl mw-content-ltr', webpage)
```
[the remark this answers is not in the document]
Comment (EhtishamSabir) [reply yt_dlp/extractor/wikimedia.py:49]: I'm getting license text in HTML format then I'm cleaning it but I did it in 2 lines in the next commit.

File: yt_dlp/extractor/wikimedia.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        license = clean_html(license)
+
+        description = get_element_by_class('description', webpage)
+        author = self._html_search_regex(r'>\s*Author\s*</td>\s*<td\b[^>]*>\s*([^<]+?)\s*</td>',
+                                         webpage, 'video author', default=None)
+        info = {'url': video_url, 'description': clean_html(description), 'ext': ext,
+                'id': video_id.replace('.' + ext, ''), 'title': self._og_search_title(webpage).replace('File:', ''),
+                'license': license, 'author': author}
```
[the remark this answers is not in the document]
Comment (EhtishamSabir) [reply yt_dlp/extractor/wikimedia.py:57]: yes thnx

**Comment by EhtishamSabir** [comment]
@pukkandan I tried to cover all of the issues can you review 1 more time, please? 

thanks.

**Comment by pukkandan** [comment]
You clearly have still not read the contributing guidelines. There are so many issues, I don't even know where to start reviewing. It was easier for me to just rewrite the whole thing.

Verify that everything is working as expected

**Comment by JimKillock** [comment]
@pukkandan I very much appreciate you doing this. 

I haven't been able to verify, but wanted to note that not all WM videos are .webm, and I wasn't sure if the code restricted the formats to just .webm in the "valid URL" part. The code in the verion by @EhtishamSabir had 
`_VALID_URL = 'https://commons.wikimedia.org/wiki/File:(?P<id>[^/]+)'`
after suggestions on YT-DL's discussion to prevent restrictions on OGG video files and so on.

**Comment by Lesmiscore** [comment]
I do believe filenames can contain slashes but question mark or hash symbol... [Also it was allowed until MediaWiki 1.27 (which is seemingly released at 2017)](https://www.mediawiki.org/wiki/Manual:$wgIllegalFileChars)

**Comment by EhtishamSabir** [comment]
> You clearly have still not read the contributing guidelines. There are so many issues, I don't even know where to start reviewing. It was easier for me to just rewrite the whole thing.
> 
> Verify that everything is working as expected
I really appreciate your edits let me run locally & verify if everything is ok.

**Comment by EhtishamSabir** [comment]
> > I haven't been able to verify, but wanted to note that not all WM videos are .webm, and I wasn't sure if the code restricted the formats to just .webm in the "valid URL" part
> 
> It can be changed to `\.\w+`

just added the default value to the license because on some pages it's not available.

**Pull request merged**